src/WebJobs.Script.WebHost/Security/KeyManagement/DefaultSecretManagerProvider.cs

```diff
@@ -45,75 +46,147 @@ public sealed class DefaultSecretManagerProvider : ISecretManagerProvider
+
+                    var keyVaultLogger = _loggerFactory.CreateLogger<KeyVaultSecretsRepository>();
+
+                    return new KeyVaultSecretsRepository(Path.Combine(_options.CurrentValue.SecretsPath, "Sentinels"), azureWebJobsSecretStorageKeyVaultUri, azureWebJobsSecretStorageKeyVaultClientId,
+                                                                        azureWebJobsSecretStorageKeyVaultClientSecret, azureWebJobsSecretStorageKeyVaultTenantId, keyVaultLogger, _environment);
+                }
+                else if (repositoryType == typeof(KubernetesSecretsRepository))
+                {
+                    repository = new KubernetesSecretsRepository(_environment, new SimpleKubernetesClient(_environment, _loggerFactory.CreateLogger<SimpleKubernetesClient>()));
+                }
+                else if (repositoryType == typeof(BlobStorageSasSecretsRepository))
+                {
+                    string secretStorageSas = _environment.GetEnvironmentVariable(EnvironmentSettingNames.AzureWebJobsSecretStorageSas);
+                    string siteSlotName = _environment.GetAzureWebsiteUniqueSlotName() ?? _hostIdProvider.GetHostIdAsync(CancellationToken.None).GetAwaiter().GetResult();
+                    repository = new BlobStorageSasSecretsRepository(Path.Combine(_options.CurrentValue.SecretsPath, "Sentinels"),
+                                                                        secretStorageSas,
+                                                                        siteSlotName,
+                                                                        _loggerFactory.CreateLogger<BlobStorageSasSecretsRepository>(),
+                                                                        _environment,
+                                                                        _azureBlobStorageProvider);
+                }
+                else if (repositoryType == typeof(BlobStorageSecretsRepository))
+                {
+                    string siteSlotName = _environment.GetAzureWebsiteUniqueSlotName() ?? _hostIdProvider.GetHostIdAsync(CancellationToken.None).GetAwaiter().GetResult();
+                    repository = new BlobStorageSecretsRepository(Path.Combine(_options.CurrentValue.SecretsPath, "Sentinels"),
+                                                                    ConnectionStringNames.Storage,
+                                                                    siteSlotName,
+                                                                    _loggerFactory.CreateLogger<BlobStorageSecretsRepository>(),
+                                                                    _environment,
+                                                                    _azureBlobStorageProvider);
+                }
+            }
+
+            if (repository == null)
```

Comment:
I am curious what it would look like if instead of throwing we log and return null here.
with this current change, it is up to callers to first check SecretsEnabled before invoking and it might easy to miss when writing new code

Reply:
Returning null here would still result in errors upstream, since when a caller accesses/uses ISecretManagerProvider.Current to manage secrets, they expect to get a working instance. Even if we were to move to a model where we created a NullSecretManager that nooped, I don't think that would be the right solution, since keys APIs and other code would return success but not actually do anything. If we're going to support "keyless", then we'll need to instrument places in the code that use keys and handle the lack appropriately. E.g. keys APIs might return an appropriate error code.

So the changes here keep the existing design where any usage of ISecretManagerProvider.Current expects that a working manager can be created/returned.

Reply:
Thank you, @mathewc . Yes, changing this behavior here would introduce the problems mentioned in addition to an API that is harder to work with and more prone to misuse.